File: FSharp/Problem0014/Program.fs
// Learn more about F# at http://fsharp.org

open System

[<EntryPoint>]
let main argv =
    let stopwatch = System.Diagnostics.Stopwatch.StartNew()
    
    let collatzSeqGenerator (num: int) = 
        match num with
            | num when num%2 = 0 -> (num/2)
            | _ -> ((3*num)+1)

    let largetCollatzSeqStartingNumber num = 
        Seq.unfold (fun x -> Some(x, (collatzSeqGenerator x))) (num)
            |> Seq.takeWhile (fun x -> x > 1)
            |> Seq.append [1]
            |> Seq.toList
            |> Seq.length

    seq { 0 .. 100 }
        |> Seq.map (fun x -> KeyValue x, largetCollatzSeqStartingNumber)
        |> Seq iter (printfn "%A" )        
        

    stopwatch.Stop()

    //printfn "Largest Prime Factor: %d (time: %fms)" largestPrimeFactor stopwatch.Elapsed.TotalMilliseconds
    0 // return an integer exit code

File: CSharp/Problem0014/Program.cs
using System;
using System.Collections.Generic;

namespace Problem0014
{
    class Program
    {
        static void Main(string[] args)
        {
            Solve();
        }

        public static void Solve()
        {
            var stopwatch = System.Diagnostics.Stopwatch.StartNew();

            var largest = new Tuple<long, long>(0L, 0L);

            for(var ind = 0L; ind < 1000000L; ind++) {
                var cLen = SeriesLen(ind);

                if(cLen > largest.Item2) {
                    largest = new Tuple<long, long>(ind, cLen);
                }
            }

            stopwatch.Stop();

            Console.WriteLine($"Largest Collatz chain is of {largest.Item2} for mnumber starting with: {largest.Item1} (time: {stopwatch.Elapsed.TotalMilliseconds}ms)");
        }

        static int SeriesLen(long num) {
            var clen = new List<long>();
            
            clen.Add(num);

            while (num > 1) {
                num = collatzNumGenerator(num);
                clen.Add(num);
            };

            return clen.Count;
        }

        static long collatzNumGenerator(long num) 
        {
            if(num%2L == 0) 
            {
                return (num/2L);
            }
            else
            {
                return ((3L*num)+1);
            }
        }
    }
}
